### ISSUE TITLE
CA1508 False positive with try finally

### ISSUE THREAD

**Issue by kevincathcart-cas** [dotnet/roslyn-analyzers#7175]
### Analyzer

**Diagnostic ID**: [CA2013](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca2013): `Do not use ReferenceEquals with value types`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.101](https://dotnet.microsoft.com/download/dotnet/5.0)
 (also verified with Microsoft.CodeAnalysis.NetAnalyzers 9.0.0-preview.24104.1, which was latest at time of submission.)

### Describe the bug

Try this minimized test case:
```c#
class CA1508
{
    class Dynamite
    {
        public void MaybeExplode()
        {
            if (Random.Shared.Next(10) > 5) throw new Exception();
        }
    }
    static void Test()
    {
        Dynamite? stick = new();
        try
        {
            stick.MaybeExplode();
            stick = null;
        }
        finally
        {
            if(stick != null)
            {
                Debug.WriteLine("This line is not dead! 50/50 chance of hitting.");
            }
        }
    }
}
```
The exact message received is:
> `warning CA1508: 'stick != null' is always 'false'. Remove or refactor the condition(s) to avoid dead code.

If I change the initialization to look like this:
```c#
Dynamite? stick = null;
stick=new();
try
{
...
```
Then the warning goes away. But remove the redundant initialization to null, and it comes back.

Now obviously that reduced testcase is not terribly realistic code.  I originally encountered the issue with code that looks more like this:

```c#
public static DeviceCollection FindAllActive(DeviceCollection deviceList)
{
    var list = new List<IDevice>();
    try
    {
        foreach (IDevice item in deviceList)
        {
            // This can potentially throw an exception, and we want to be sure to clean up if that happens
            if(IsDeviceActive(item))
                list.Add(item.Clone());
        }

        // DeviceCollection assumes ownership of these disposables, and the caller will assume ownership of this return value.
        var collection = new DeviceCollection(list);
        list = null;
        return collection;
    }
    finally
    {
        // If probing or creating the DeviceCollection failed, we still own the cloned Devices, and must dispose of them.
        if (list != null)
        {
            foreach (IDevice item in list)
            {
                item.Dispose();
            }
        }
    }
}
```
